### PR TITLE
complete alter table partition grammar

### LINKED ISSUE #10409: Complete oracle partition grammar
relate #9694.

support oracle alter table partition grammar

image

we need support 
- modify_table_default_attrs
- alter_automatic_partitioning
- alter_interval_partitioning
- set_subpartition_template
- modify_table_partition
- modify_table_subpartition
- move_table_partition
- move_table_subpartition
- add_external_partition_attrs
- add_table_partition
- coalesce_table_partition
- drop_external_partition_attrs
- drop_table_partition
- drop_table_subpartition
- rename_partition_subpart
- truncate_partition_subpart
- split_table_partition
- split_table_subpartition
- merge_table_partitions
- merge_table_subpartitions
- exchange_partition_subpart

doc:
https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/ALTER-TABLE.html#GUID-552E7373-BF93-477D-9DA3-B2C9386F2877

### PR DESCRIPTION
**Pull request by zhujunxxxxx**
Fixes #10409

Changes proposed in this pull request:
- support `modifyTablePartition` `moveTablePartition` `coalesceTablePartition` grammar

### CODE REVIEW

**Comment by zhujunxxxxx** [comment]
@tristaZero Please review this PR.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4 [hunk 1]
```diff
@@ -964,23 +964,136 @@ alterSynonym
...
+        |  (PARTITION partitionName? addHashPartitionClause (COMMA_ PARTITION partitionName? addHashPartitionClause)*)
+        ) dependentTablesClause?
     ;
 
 addRangePartitionClause
-    : PARTITION partitionName? rangeValuesClause tablePartitionDescription
+    : rangeValuesClause tablePartitionDescription? externalPartSubpartDataProps?
     ((LP_? rangeSubpartitionDesc (COMMA_ rangeSubpartitionDesc)* | listSubpartitionDesc (COMMA_ listSubpartitionDesc)* | individualHashSubparts (COMMA_ individualHashSubparts)* RP_?)
-        | hashSubpartitionQuantity)?
+    | hashSubpartitionQuantity)? updateIndexClauses?
```
Comment (Liangda-w) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4:1082]: `hashSubpartitionQuantity` is not the same as `hash_subparts_by_quantity`, please have a check:)

**Review by Liangda-w — CHANGES_REQUESTED** [review]

**Comment by tristaZero** [comment]
@zhujunxxxxx  Looks @Liangda-w gave a detailed review.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4 [hunk 2]
```diff
@@ -964,23 +964,136 @@ alterSynonym
...
+    | PARTITION FOR LR_ partitionKeyValue (COMMA_ partitionKeyValue)* RP_
+    ;
+
+addRangeSubpartition
+    : ADD rangeSubpartitionDesc (COMMA_ rangeSubpartitionDesc)* dependentTablesClause? updateIndexClauses?
+    ;
+
+dependentTablesClause
+    : DEPENDENT TABLES LP_ tableName LP_ partitionSpec (COMMA_ partitionSpec)* RP_
+    (tableName LP_ partitionSpec (COMMA_ partitionSpec)* RP_)* RP_
```
[the remark this answers is not in the document]
Comment (zhujunxxxxx) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4:1010]: yes, it was missed.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4 (hunk 1 above)
Comment (zhujunxxxxx) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4:1082]: Yes, I will fix it.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4 [hunk 3]
```diff
@@ -964,23 +964,136 @@ alterSynonym
...
+    ;
+
+partitionSpec
+    : PARTITION partitionName? tablePartitionDescription?
+    ;
+
+partitionAttributes
+    : (physicalAttributesClause | loggingClause | allocateExtentClause | deallocateUnusedClause | shrinkClause)?
+      (OVERFLOW (physicalAttributesClause | loggingClause | allocateExtentClause | deallocateUnusedClause)?)?
+      tableCompression? inmemoryClause?
```
[the remark this answers is not in the document]
Comment (zhujunxxxxx) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4:1049]: ok, next PR fix it.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4 [hunk 4]
```diff
@@ -964,23 +964,142 @@ alterSynonym
...
+    | PARTITION FOR LR_ partitionKeyValue (COMMA_ partitionKeyValue)* RP_
+    ;
+
+addRangeSubpartition
+    : ADD rangeSubpartitionDesc (COMMA_ rangeSubpartitionDesc)* dependentTablesClause? updateIndexClauses?
+    ;
+
+dependentTablesClause
+    : DEPENDENT TABLES LP_ tableName LP_ partitionSpec (COMMA_ partitionSpec)* RP_
+    (tableName LP_ partitionSpec (COMMA_ partitionSpec)* RP_)* RP_
```
Comment (Liangda-w) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4:1010]: @zhujunxxxxx Thanks for all the fix, looks good! 
Only one point here: did you eventually forget to delete the old line?

**Comment by tristaZero** [comment]
Hi @zhujunxxxxx @Liangda-w  it is the right time to merge this one?

**Comment by Liangda-w** [comment]
> Hi @zhujunxxxxx @Liangda-w it is the right time to merge this one?

yes, I think so:)

**Comment by zhujunxxxxx** [comment]
@Liangda-w thank for u review.

**Comment by zhujunxxxxx** [comment]
@Liangda-w Can you give me a contact number, maybe we have a chance to work together in the future.

**Review by tristaZero — APPROVED** [review]

**Comment by Liangda-w** [comment]
@zhujunxxxxx ur welcome! Sure, you can contact me via liangda_wang@outlook.de

**Pull request merged**